lib/services/livesync/livesync-service.ts

```diff
@@ -1,136 +1,443 @@
+import { hook } from "../../common/helpers";
+import { APP_FOLDER_NAME, PACKAGE_JSON_FILE_NAME, LiveSyncTrackActionNames } from "../../constants";
+import { FileExtensions, DeviceTypes } from "../../common/constants";
 
-let choki = require("chokidar");
+const LiveSyncEvents = {
+	liveSyncStopped: "liveSyncStopped",
+	// In case we name it error, EventEmitter expects instance of Error to be raised and will also raise uncaught exception in case there's no handler
+	liveSyncError: "liveSyncError",
+	liveSyncExecuted: "liveSyncExecuted",
+	liveSyncStarted: "liveSyncStarted",
+	liveSyncNotification: "notify"
+};
 
-class LiveSyncService implements ILiveSyncService {
-	private _isInitialized = false;
+// TODO: emit events for "successfull livesync", "stoppedLivesync",
+export class LiveSyncService extends EventEmitter implements ILiveSyncService {
+	// key is projectDir
+	private liveSyncProcessesInfo: IDictionary<ILiveSyncProcessInfo> = {};
 
-	constructor(private $errors: IErrors,
-		private $platformsData: IPlatformsData,
-		private $platformService: IPlatformService,
-		private $injector: IInjector,
-		private $devicesService: Mobile.IDevicesService,
-		private $options: IOptions,
-		private $logger: ILogger,
-		private $dispatcher: IFutureDispatcher,
-		private $hooksService: IHooksService,
+	constructor(protected $platformService: IPlatformService,
+		private $projectDataService: IProjectDataService,
+		protected $devicesService: Mobile.IDevicesService,
+		private $mobileHelper: Mobile.IMobileHelper,
+		protected $devicePlatformsConstants: Mobile.IDevicePlatformsConstants,
+		private $nodeModulesDependenciesBuilder: INodeModulesDependenciesBuilder,
+		protected $logger: ILogger,
 		private $processService: IProcessService,
-		private $nodeModulesDependenciesBuilder: INodeModulesDependenciesBuilder) { }
-
-	public get isInitialized(): boolean { // This function is used from https://github.com/NativeScript/nativescript-dev-typescript/blob/master/lib/before-prepare.js#L4
-		return this._isInitialized;
+		private $hooksService: IHooksService,
+		protected $injector: IInjector) {
+		super();
 	}
 
-	public async liveSync(platform: string, projectData: IProjectData, applicationReloadAction?: (deviceAppData: Mobile.IDeviceAppData) => Promise<void>): Promise<void> {
-		if (this.$options.justlaunch) {
-			this.$options.watch = false;
+	@hook("liveSync")
+	public async liveSync(deviceDescriptors: ILiveSyncDeviceInfo[],
+		liveSyncData: ILiveSyncInfo): Promise<void> {
+		// TODO: Initialize devicesService before that.
+		const projectData = this.$projectDataService.getProjectData(liveSyncData.projectDir);
+		// In case liveSync is called for a second time for the same projectDir.
+		const isAlreadyLiveSyncing = this.liveSyncProcessesInfo[projectData.projectDir] && !this.liveSyncProcessesInfo[projectData.projectDir].isStopped;
+		this.setLiveSyncProcessInfo(liveSyncData.projectDir, deviceDescriptors);
+
+		// TODO: Check if the _.difference actually works.
```

Comment:
We have and it does. I am actually a bit confused about the existence of this comment as a whole